README.md
# touch_interceptor

A widget that intercepts touch events and sends them to other widgets underneath.

![Schema](https://raw.githubusercontent.com/parsodyl/touch_interceptor/master/touch_interceptor.png?sanitize=true)


### Useful when:

- there is an opaque layer over a widget, and you want to make it transparent to touch events 
- you want seamlessly transfer a touch event from one widget to another

### Syntax:

```dart
// interceptor
TouchInterceptor(
  // other component(s)
  child: OtherWidget(
    // consumer
    child: TouchConsumer(
      onTouchDown: () {
        // do something
      },
      onTouchUp: () {
        // do something
      },
      onTouchEnter: () {
        // do something
      },
      onTouchExit: () {
        // do something
      },
    ),
  ),
),
```

### See the examples:
- [Basic example](https://github.com/parsodyl/touch_interceptor/tree/master/example)
- [Frozen layer](https://github.com/parsodyl/touch_interceptor_examples/tree/main/frozen_layer)
- [Colorful pads](https://github.com/parsodyl/touch_interceptor_examples/tree/main/colorful_pads)


CHANGELOG.md
## [0.1.1]

* Updated documentation.

## [0.1.0]

* Null safety migration.

## [0.0.2]

* Key generation fixed.
* Null returned value prevented. 

## [0.0.1]

* First release.


test/ti_basic_test.dart
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:touch_interceptor/touch_interceptor.dart';

void main() {
  testWidgets(
    'TouchInterceptor is added to the tree',
    (tester) async {
      const touchInterceptor = TouchInterceptor();
      const testingWidget = TestingWidget(child: touchInterceptor);
      // pump
      await tester.pumpWidget(testingWidget);
      // check
      expect(find.byWidget(touchInterceptor), findsOneWidget);
    },
  );

  testWidgets(
    'A TouchConsumer is placed underneath a TouchInterceptor.',
    (tester) async {
      const touchConsumer = TouchConsumer();
      const touchInterceptor = TouchInterceptor(child: touchConsumer);
      const testingWidget = TestingWidget(child: touchInterceptor);
      // pump
      await tester.pumpWidget(testingWidget);
      // check
      final touchConsumerFinder = find.byWidget(touchConsumer);
      expect(
        touchConsumerFinder,
        findsOneWidget,
      );
      expect(
        find.descendant(
          of: find.byWidget(touchInterceptor),
          matching: touchConsumerFinder,
        ),
        findsOneWidget,
      );
    },
  );

  testWidgets(
    'A TouchConsumer is not placed underneath a TouchInterceptor.',
    (tester) async {
      const touchConsumer = TouchConsumer();
      // try to insert
      await tester.pumpWidget(const TestingWidget(child: touchConsumer));
      // check
      expect(
        tester.takeException(),
        isA<TouchInterceptorNotFoundError>()
            .having((err) => err.toString(), 'message', isNotNull),
      );
    },
  );
}

class TestingWidget extends StatelessWidget {
  const TestingWidget({
    Key? key,
    required this.child,
  }) : super(key: key);

  final Widget child;

  @override
  Widget build(BuildContext context) {
    return child;
  }
}


test/ti_gestures_test.dart
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:touch_interceptor/touch_interceptor.dart';

void main() {
  testWidgets(
    'TouchInterceptor receives pointer events (up and then down)',
    (tester) async {
      const testingWidget = TestApplicationWidget();
      // pump
      await tester.pumpWidget(testingWidget);
      // check
      expect(find.text('none'), findsOneWidget);
      // press the widget (down)
      final gesture = await tester.press(find.byKey(const ValueKey('TI')));
      // pump again
      await tester.pumpAndSettle();
      // check
      expect(find.text('down'), findsOneWidget);
      // release from press (up)
      await gesture.up();
      // pump again
      await tester.pumpAndSettle();
      // check
      expect(find.text('up'), findsOneWidget);
    },
  );

  testWidgets(
    'TouchInterceptor receives pointer events (enter and then up)',
    (tester) async {
      const testingWidget = TestApplicationWidget();
      // pump
      await tester.pumpWidget(testingWidget);
      // check
      expect(find.text('none'), findsOneWidget);
      // move the pointer in
      final gesture = await tester.startGesture(Offset.zero);
      const position = Offset(
        TestApplicationWidget.padding,
        TestApplicationWidget.padding,
      );
      await gesture.moveTo(position);
      // pump again
      await tester.pumpAndSettle();
      // check
      expect(find.text('enter'), findsOneWidget);
      // release from press (up)
      await gesture.up();
      // pump again
      await tester.pumpAndSettle();
      // check
      expect(find.text('up'), findsOneWidget);
    },
  );

  testWidgets(
    'TouchInterceptor receives pointer events (enter and then exit)',
    (tester) async {
      const testingWidget = TestApplicationWidget();
      // pump
      await tester.pumpWidget(testingWidget);
      // check
      expect(find.text('none'), findsOneWidget);
      // move the pointer in
      final gesture = await tester.startGesture(Offset.zero);
      const position = Offset(
        TestApplicationWidget.padding,
        TestApplicationWidget.padding,
      );
      await gesture.moveTo(position);
      // pump again
      await tester.pumpAndSettle();
      // check
      expect(find.text('enter'), findsOneWidget);
      // move the pointer out
      await gesture.moveTo(-position);
      // pump again
      await tester.pumpAndSettle();
      // check
      expect(find.text('exit'), findsOneWidget);
    },
  );

  testWidgets(
    'TouchInterceptor receives pointer events (enter and then cancel)',
    (tester) async {
      const testingWidget = TestApplicationWidget();
      // pump
      await tester.pumpWidget(testingWidget);
      // check
      expect(find.text('none'), findsOneWidget);
      // move the pointer in
      final gesture = await tester.startGesture(Offset.zero);
      const position = Offset(
        TestApplicationWidget.padding,
        TestApplicationWidget.padding,
      );
      await gesture.moveTo(position);
      // pump again
      await tester.pumpAndSettle();
      // check
      expect(find.text('enter'), findsOneWidget);
      // cancel the gesture
      await gesture.cancel();
      // pump again
      await tester.pumpAndSettle();
      // check
      expect(find.text('exit'), findsOneWidget);
    },
  );
}

class TestApplicationWidget extends StatefulWidget {
  const TestApplicationWidget({Key? key}) : super(key: key);

  static const padding = 16.0;

  @override
  State<TestApplicationWidget> createState() => _TestApplicationWidgetState();
}

class _TestApplicationWidgetState extends State<TestApplicationWidget> {
  String _consumerEvent = 'none';

  @override
  Widget build(BuildContext context) {
    return TouchInterceptor(
      key: const ValueKey('TI'),
      child: Padding(
        padding: const EdgeInsets.all(TestApplicationWidget.padding),
        child: TouchConsumer(
          key: const ValueKey('TC'),
          onTouchDown: () => setState(() => _consumerEvent = 'down'),
          onTouchUp: () => setState(() => _consumerEvent = 'up'),
          onTouchEnter: () => setState(() => _consumerEvent = 'enter'),
          onTouchExit: () => setState(() => _consumerEvent = 'exit'),
          child: Text(
            '$_consumerEvent',
            textDirection: TextDirection.ltr,
          ),
        ),
      ),
    );
  }
}


example/test/widget_test.dart
import 'package:flutter_test/flutter_test.dart';


void main() {
  testWidgets('smoke test', (WidgetTester tester) async {
    // test implementation
  });
}


lib/src/touch_interceptor.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/rendering.dart';
import 'package:flutter/widgets.dart';

/// A widget that intercepts pointer events and sends them to every
/// [TouchConsumer] widget placed underneath.
///
/// ## Layout behavior
///
/// _See [BoxConstraints] for an introduction to box layout models._
///
/// If it has a child, this widget defers to the child for sizing behavior. If
/// it does not have a child, it grows to fit the parent instead.
class TouchInterceptor extends StatefulWidget {
  /// Creates a [TouchInterceptor] widget.
  const TouchInterceptor({Key? key, this.child}) : super(key: key);

  /// The [child] contained by the interceptor.
  ///
  /// {@macro flutter.widgets.child}
  final Widget? child;

  @override
  _TouchInterceptorState createState() => _TouchInterceptorState();
}

class _TouchInterceptorState extends State<TouchInterceptor> {
  final _keySet = <_TouchKey>{};

  @override
  Widget build(BuildContext context) {
    return _KeyRegister(
      keySet: _keySet,
      child: Listener(
        onPointerDown: (PointerDownEvent details) =>
            _sendAction(_TouchAction.down, details.position),
        onPointerMove: (PointerMoveEvent details) =>
            _sendAction(_TouchAction.move, details.position),
        onPointerUp: (PointerUpEvent details) =>
            _sendAction(_TouchAction.up, details.position),
        onPointerCancel: (PointerCancelEvent details) =>
            _sendAction(_TouchAction.cancel, details.position),
        child: AbsorbPointer(child: widget.child),
      ),
    );
  }

  void _sendAction(_TouchAction touchAction, Offset offset) {
    for (_TouchKey key in _keySet) {
      key.currentState!.dispatchTouch(offset, touchAction);
    }
  }

  @override
  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
    super.debugFillProperties(properties);

    properties.add(IterableProperty<String>(
        'keys', _keySet.map((_TouchKey key) => key.toString()).toList(),
        ifEmpty: '<none>'));
  }
}

/// A widget that listens for events from the nearest [TouchInterceptor]
/// ancestor and calls callbacks in response to them.
///
/// ## Layout behavior
///
/// _See [BoxConstraints] for an introduction to box layout models._
///
/// If it has a child, this widget defers to the child for sizing behavior. If
/// it does not have a child, it grows to fit the parent instead.
class TouchConsumer extends StatefulWidget {
  // Creates a [TouchConsumer] widget.
  const TouchConsumer({
    Key? key,
    this.onTouchDown,
    this.onTouchEnter,
    this.onTouchExit,
    this.onTouchUp,
    this.child,
  }) : super(key: key);

  /// Called when a pointer comes into contact with the screen at this widget's
  /// location.
  final VoidCallback? onTouchDown;

  /// Called when a pointer that has previously come into contact with the screen
  /// changes position and reaches this widget's location.
  final VoidCallback? onTouchEnter;

  /// Called when a pointer that has previously come into contact with the screen
  /// changes position and leaves this widget's location.
  final VoidCallback? onTouchExit;

  /// Called when a pointer that has previously come into contact with the screen
  /// stops being in contact with the screen at this widget's location.
  final VoidCallback? onTouchUp;

  /// The widget below this widget in the tree.
  ///
  /// {@macro flutter.widgets.child}
  final Widget? child;

  @override
  _TouchConsumerState createState() => _TouchConsumerState();
}

class _TouchConsumerState extends State<TouchConsumer> {
  _TouchKey? _key;

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();

    _registerKeyIfIsNot();
  }

  @override
  void deactivate() {
    _unregisterKey();

    super.deactivate();
  }

  @override
  Widget build(BuildContext context) {
    final callbacks = _TouchCallbacks(
      widget.onTouchDown,
      widget.onTouchEnter,
      widget.onTouchExit,
      widget.onTouchUp,
    );
    return _TouchConsumerCore(
      key: _key,
      child: widget.child,
      callbacks: callbacks,
    );
  }

  void _registerKeyIfIsNot() {
    if (_key != null && _keyRegister.isKeyRegistered(_key)) {
      return;
    }
    _key = _TouchKey(); // keep it final!
    _keyRegister.registerNewKey(_key!);
  }

  void _unregisterKey() {
    _keyRegister.unregisterKey(_key);
    _key = null;
  }

  _KeyRegister get _keyRegister {
    final keyRegister = _KeyRegister.of(context);
    if (keyRegister == null) {
      throw TouchInterceptorNotFoundError._();
    }
    return keyRegister;
  }
}

/// Thrown when a TouchInterceptor is not found above in the widget tree.
class TouchInterceptorNotFoundError extends Error {
  TouchInterceptorNotFoundError._();

  @override
  String toString() {
    return 'Error: Could not find a TouchInterceptor widget above this TouchConsumer widget.';
  }
}

enum _TouchAction { down, move, up, cancel }

class _KeyRegister extends InheritedWidget {
  const _KeyRegister({
    Key? key,
    required Set<_TouchKey> keySet,
    required Widget child,
  })  : _keySet = keySet,
        super(key: key, child: child);

  final Set<_TouchKey> _keySet;

  void registerNewKey(_TouchKey newKey) {
    _keySet.add(newKey);
  }

  bool isKeyRegistered(_TouchKey? key) {
    return _keySet.contains(key);
  }

  void unregisterKey(_TouchKey? key) {
    _keySet.remove(key);
  }

  List<_TouchKey> get keys => _keySet.toList(growable: false);

  @override
  bool updateShouldNotify(_KeyRegister oldWidget) {
    return _keySet != oldWidget._keySet;
  }

  static _KeyRegister? of(BuildContext context) =>
      context.dependOnInheritedWidgetOfExactType<_KeyRegister>();
}

class _TouchConsumerCore extends StatefulWidget {
  const _TouchConsumerCore({
    required _TouchKey? key,
    required this.callbacks,
    required this.child,
  }) : super(key: key);

  final _TouchCallbacks callbacks;
  final Widget? child;

  @override
  _TouchConsumerCoreState createState() => _TouchConsumerCoreState();
}

class _TouchConsumerCoreState extends State<_TouchConsumerCore> {
  bool _hasTouchEntered = false;

  @override
  Widget build(BuildContext context) {
    return Container(child: widget.child);
  }

  void dispatchTouch(Offset touchPosition, _TouchAction action) {
    final hit = _hitTest(touchPosition);
    if (hit) {
      switch (action) {
        case _TouchAction.down:
          if (!_hasTouchEntered) {
            setState(() {
              _hasTouchEntered = true;
            });
            widget.callbacks.down();
          }
          break;
        case _TouchAction.move:
          if (!_hasTouchEntered) {
            setState(() {
              _hasTouchEntered = true;
            });
            widget.callbacks.enter();
          }
          break;
        case _TouchAction.up:
          if (_hasTouchEntered) {
            setState(() {
              _hasTouchEntered = false;
            });
            widget.callbacks.up();
          }
          break;
        case _TouchAction.cancel:
          if (_hasTouchEntered) {
            setState(() {
              _hasTouchEntered = false;
            });
            widget.callbacks.exit();
          }
          break;
      }
    } else {
      if (_hasTouchEntered) {
        setState(() {
          _hasTouchEntered = false;
        });
        widget.callbacks.exit();
      }
    }
  }

  bool _hitTest(Offset coordinates) {
    final renderBox = context.findRenderObject() as RenderBox;
    final localPosition = renderBox.globalToLocal(coordinates);
    final result = BoxHitTestResult();
    return renderBox.hitTest(result, position: localPosition);
  }
}

class _TouchKey extends GlobalKey<_TouchConsumerCoreState> {
  const _TouchKey() : super.constructor();

  @override
  String toString() {
    return '[TK<$hashCode>]';
  }
}

@immutable
class _TouchCallbacks {
  const _TouchCallbacks(
    this._onTouchDown,
    this._onTouchEnter,
    this._onTouchExit,
    this._onTouchUp,
  );

  final VoidCallback? _onTouchDown;
  final VoidCallback? _onTouchEnter;
  final VoidCallback? _onTouchExit;
  final VoidCallback? _onTouchUp;

  void down() => _onTouchDown?.call();

  void enter() => _onTouchEnter?.call();

  void exit() => _onTouchExit?.call();

  void up() => _onTouchUp?.call();
}


example/touch_interceptor_example.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/rendering.dart';
import 'package:touch_interceptor/touch_interceptor.dart';

void main() {
  WidgetsFlutterBinding.ensureInitialized();
  runApp(ExampleApp());
}

class ExampleApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return CupertinoApp(
      theme: const CupertinoThemeData(
        brightness: Brightness.dark,
      ),
      home: DemoView(),
    );
  }
}

class DemoView extends StatefulWidget {
  @override
  _DemoViewState createState() => _DemoViewState();
}

class _DemoViewState extends State<DemoView> {
  Color _shapeColor = CupertinoColors.destructiveRed;

  var _lastEvent = '';

  @override
  Widget build(BuildContext context) {
    return SafeArea(
      child: Padding(
        padding: const EdgeInsets.symmetric(
          horizontal: 16.0,
        ),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: <Widget>[
            Expanded(
              child: Center(
                child: TouchInterceptor(
                  child: Container(
                    decoration: BoxDecoration(
                      border: Border.all(
                        color: _shapeColor,
                        width: 4.0,
                      ),
                    ),
                    child: Padding(
                      padding: const EdgeInsets.all(60.0),
                      child: TouchConsumer(
                        onTouchDown: () {
                          setState(() {
                            _shapeColor = CupertinoColors.activeGreen;
                            _lastEvent = 'down!';
                          });
                        },
                        onTouchUp: () {
                          setState(() {
                            _shapeColor = CupertinoColors.white;
                            _lastEvent = 'up!';
                          });
                        },
                        onTouchEnter: () {
                          setState(() {
                            _shapeColor = CupertinoColors.activeGreen;
                            _lastEvent = 'hover!';
                          });
                        },
                        onTouchExit: () {
                          setState(() {
                            _shapeColor = CupertinoColors.destructiveRed;
                            _lastEvent = 'away!';
                          });
                        },
                        child: Container(
                          decoration: BoxDecoration(
                            color: _shapeColor,
                            shape: BoxShape.circle,
                          ),
                          child: const SizedBox(width: 80, height: 80),
                        ),
                      ),
                    ),
                  ),
                ),
              ),
            ),
            const SizedBox(height: 45),
            Text('Touch: $_lastEvent'),
            const SizedBox(height: 5),
          ],
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/cupertino.dart';
import 'package:touch_interceptor/touch_interceptor.dart';

void main() {
  WidgetsFlutterBinding.ensureInitialized();
  runApp(const ExampleApp());
}

class ExampleApp extends StatelessWidget {
  const ExampleApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const CupertinoApp(
      theme: CupertinoThemeData(
        brightness: Brightness.dark,
      ),
      home: DemoView(),
    );
  }
}

class DemoView extends StatefulWidget {
  const DemoView({Key? key}) : super(key: key);

  @override
  _DemoViewState createState() => _DemoViewState();
}

class _DemoViewState extends State<DemoView> {
  static const _tiColor = CupertinoColors.white;
  Color _tcColor = CupertinoColors.white;

  var _lastEvent = 'none';

  @override
  Widget build(BuildContext context) {
    return SafeArea(
      child: Padding(
        padding: const EdgeInsets.symmetric(horizontal: 16.0),
        child: Column(
          children: <Widget>[
            Expanded(
              child: Center(
                child: Column(
                  mainAxisSize: MainAxisSize.min,
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    const Text(
                      'TouchInterceptor',
                      style: TextStyle(
                        color: _tiColor,
                        fontWeight: FontWeight.bold,
                      ),
                    ),
                    TouchInterceptor(
                      child: DecoratedBox(
                        decoration: BoxDecoration(
                          borderRadius: BorderRadius.circular(16),
                          border: Border.all(
                            color: _tiColor,
                            width: 4.0,
                          ),
                        ),
                        child: Padding(
                          padding: const EdgeInsets.all(90.0),
                          child: Column(
                            mainAxisSize: MainAxisSize.min,
                            crossAxisAlignment: CrossAxisAlignment.start,
                            children: [
                              Text(
                                'TouchConsumer',
                                style: TextStyle(
                                  color: _tcColor,
                                  fontWeight: FontWeight.bold,
                                ),
                              ),
                              TouchConsumer(
                                onTouchDown: () => setState(() {
                                  _tcColor = CupertinoColors.activeBlue;
                                  _lastEvent = 'down';
                                }),
                                onTouchUp: () => setState(() {
                                  _tcColor = CupertinoColors.activeOrange;
                                  _lastEvent = 'up';
                                }),
                                onTouchEnter: () => setState(() {
                                  _tcColor = CupertinoColors.activeGreen;
                                  _lastEvent = 'enter';
                                }),
                                onTouchExit: () => setState(() {
                                  _tcColor = CupertinoColors.destructiveRed;
                                  _lastEvent = 'exit';
                                }),
                                child: DecoratedBox(
                                  decoration: BoxDecoration(
                                    color: _tcColor,
                                    borderRadius: BorderRadius.circular(16),
                                  ),
                                  child: const SizedBox.square(dimension: 140),
                                ),
                              ),
                            ],
                          ),
                        ),
                      ),
                    ),
                  ],
                ),
              ),
            ),
            const SizedBox(height: 8),
            Text.rich(
              TextSpan(
                text: 'Tap or click inside the box to change color!\n',
                children: [
                  const TextSpan(text: 'Last event: '),
                  TextSpan(
                    text: _lastEvent,
                    style: TextStyle(color: _tcColor),
                  ),
                ],
              ),
            ),
            const SizedBox(height: 8),
          ],
        ),
      ),
    );
  }
}
